CONTRIBUTING.md

```diff
@@ -0,0 +1,277 @@
+
+### Formatting
+
+The project uses rustfmt with custom configuration (`rustfmt.toml`):
+
+```bash
+# Check formatting
+cargo fmt -- --check
+
+# Apply formatting
+cargo fmt
+```
+
+### Linting
+
+All clippy warnings are treated as errors in CI:
+
+```bash
+# Lint library crates
+cargo clippy --workspace --lib -- -D clippy::all
+
+# Or exclude the main binary explicitly
+cargo clippy --workspace --exclude muffinos -- -D clippy::all
+```
+
+**Important:** Always run clippy on library crates with `--lib` to avoid bare-metal compilation issues.
+
+## Development Workflow
+
+### Before Submitting a PR
+
+Run these commands in order to validate your changes:
+
+```bash
+# 1. Format check (fastest)
+cargo fmt -- --check
+
+# 2. Lint check
+cargo clippy --workspace --lib -- -D clippy::all
+
+# 3. Build check
+cargo build --workspace --lib
+
+# 4. Test modified crates
+cargo test -p <modified_crate>
+
+# 5. (Optional) Miri tests if you changed kernel crates
+cargo miri setup
+cargo miri test -p <modified_crate>
+
+# 6. (Optional) Full build
+cargo build --release
+```
+
+### CI Pipeline
+
+GitHub Actions runs on every push with these jobs:
+
+1. **Lint:** Checks formatting and runs clippy with `-D clippy::all`
+2. **Test:** Runs tests in both debug and release modes
+3. **Miri:** Tests each kernel crate with Miri for undefined behavior
+4. **Build:** Creates the bootable ISO and uploads artifacts
+
+The CI also runs twice daily on a schedule.
+
+## Running the OS
+
+To build and run Muffin OS in QEMU:
+
+```bash
+# Run with default settings
+cargo run
+
+# Run without GUI
+cargo run -- --headless
+
+# Run with GDB debugging (connects on localhost:1234)
+cargo run -- --debug
+
+# Customize CPU cores and memory
+cargo run -- --smp 4 --mem 512M
+
+# Build ISO without running
+cargo run -- --no-run
+```
+
+## Project Guidelines
+
+### Code Style
+
+- Follow Rust naming conventions and idioms
+- Keep functions focused and modular
+- Document public APIs with doc comments
+- Use descriptive variable names
+- Prefer safe Rust; justify all `unsafe` blocks with safety comments
+
+### Commit Messages
+
+- Use clear, descriptive commit messages
+- Start with a verb in present tense (e.g., "Add", "Fix", "Update")
+- Reference issue numbers when applicable
+
+### Pull Requests
+
+- Keep PRs focused on a single feature or fix
+- Update documentation for user-facing changes
+- Ensure all CI checks pass
+- Add tests when adding testable functionality to crates
+
+## License
+
+Muffin OS is dual-licensed under Apache-2.0 OR MIT. All contributions must be compatible with this licensing.
+
+## Getting Help
+
+- Check existing issues for similar problems
+- Review the CI logs for detailed error messages
+- Ask questions in issue discussions
+
+## Additional Notes
+
+### Known Limitations
+
+- The kernel binary uses a custom linker script and cannot run standard Rust tests
+- Some kernel structures have intentional dead code warnings for fields used by hardware
```

Comment:
This shouldn't be the case, and if it is, this is a bug. Please remove this line in either case.